Add Wave API changes (#25)
seqeralabs/docs: wave_docs/api.mdx

This page summarizes the API provided by the Wave container service.

> ## POST `/container-token`

Reviewer: This endpoint should be marked as *deprecated*

| ----------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| `containerImage`                    | Name of the container to be served, e.g., `docker.io/library/ubuntu:latest` (optional). If omitted, the `containerFile` must be provided.          |
| `containerConfig.entrypoint`        | The container entrypoint command e.g. `['/bin/bash']`                                                                                              |
| `containerConfig.cmd`               | The launch command to be used by the Wave container, e.g., `['echo', 'Hello world']` (optional).                                                   |
| `containerConfig.env`               | The environment variables to be defined in the Wave container e.g. `['FOO=one','BAR=two']` (optional).                                             |
| `containerConfig.workingDir`        | The work directory to be used in the Wave container e.g. `/some/work/dir` (optional).                                                              |

> | `containerConfig.layers.location`   | Either a HTTP URL from where a container layer tar gzip file should be downloaded or tar gzip file base64 encoded and prefixed with `data:` string (optional). |

Author: @munishchouhan, can you elaborate on what this field does? Thanks!

> | `containerConfig.layers.location`   | Either a HTTP URL from where a container layer tar gzip file should be downloaded or tar gzip file base64 encoded and prefixed with `data:` string (optional). |

Reviewer 2: This specifies the location of container image layer, which is a tar file

> | `containerConfig.layers.tarDigest`  | The SHA256checksum of the provided tar file e.g. `sha256:a7c724b02...`.                                                                            |

Reviewer 3: ```suggestion
| `containerConfig.layers.tarDigest`  | The SHA256checksum of the provided tar file, e.g., `sha256:a7c724b02...`.                                                                            |
```

| `spackFile`                         | Spack recipe file encoded as base64 string.                                                                                                        |
| `containerPlatform`                 | Target container architecture of the built container e.g. `linux/amd64` (optional). Currently only supporting amd64 and arm64.                     |
| `buildRepository`                   | Container repository where container builds should be pushed e.g. `docker.io/user/my-image` (optional).                                            |
| `cacheRepository`                   | Container repository used to cache build layers `docker.io/user/my-cache` (optional).                                                              |
| `timestamp`                         | Request submission timestap using ISO-8601.                                                                                                        |
| `fingerprint`                       | Request unique fingerprint.                                                                                                                        |

> | `freeze`                            | The container provisioned will be stored in the specified repository in a permanently.                                                             |

Reviewer 3: Incomplete sentence.

Author: @munishchouhan, can you elaborate on what `freeze` is used for here? Thanks!

Reviewer 2: Freeze requires `buildRepository` to push the build container to a user-defined repository. This will provide the container URL from the user-defined repository, not the generated URL. This URL won't change with simultaneous calls to wave.

> | `towerEndpoint`                     | Seqera Platform service endpoint from where container repositories credentials are retrieved (optional). Default `https://api.cloud.seqera.io`.                       |

Reviewer 3: ```suggestion
| `towerEndpoint`                     | Seqera Platform service endpoint from where container repository credentials are retrieved (optional). Default `https://api.cloud.seqera.io`.                       |
```

> | `towerEndpoint`                     | Seqera Platform service endpoint from where container repositories credentials are retrieved (optional). Default `https://api.cloud.seqera.io`.                       |

Reviewer 3: We call these container registries in Seqera Platform. If that is the correct term here, let's be consistent with what Platform docs and UI already use.

Author: @llewellyn-sl, I'll ask and see if there is any particular reason to use different terminology.

> | `towerAccessToken`                  | Access token of the user account granting the access to the Seqera Platform service specified via `towerEndpoint` (optional).                      |

Reviewer 3: ```suggestion
| `towerAccessToken`                  | Access token of the user account granting access to the Seqera Platform service specified via `towerEndpoint` (optional).                      |
```

> | `towerWorkspaceId`                  | ID of the Seqera Platform workspace from where the container repositories credentials are retrieved (optional). When omitted the personal workspace is used. |

Reviewer 3: ```suggestion
| `towerWorkspaceId`                  | ID of the Seqera Platform workspace from where the container repository credentials are retrieved (optional). When omitted, the personal workspace is used. |
```

        env: string[],
        workingDir: string
        layers: [
            {
                location: string,
                gzipDigest: string,
                gzipSize: string,
                tarDigest: string
            },
            ...
        ]
    },
    condaFile: string,
    spackFile: string,
    containerPlatform: string,
    buildRepository: string,
    cacheRepository: string,
    timestamp: string,
    fingerprint: string,
    freeze: boolean,
    towerAccessToken: string,
    towerRefreshToken: string,
    towerEndpoint: string,
    towerWorkspaceId: number,
    packages: {
                  type: string,
                  packages: string[],
                  channels: string[],
                  envFile: string,
                  condaOpts: {
                                mambaImage: string,
                                commands: string[],
                                basePackages: string
                              }
                  spackOpts:{
                                commands: string[],
                                basePackages: string
                            }                     
                  
              }
}
```

> Note: You can read the description of al attributes except packages from [here](#container-token-request-attributes)

Reviewer 3: Admonition?

all* attributes, except packages from...

Author: @munishchouhan, what are the container token request attributes? Thanks!

| Attribute      | Description                                                                 |
|----------------|-----------------------------------------------------------------------------|

> | `packages`     | this object specifies conda or spack packages environment information       |

Reviewer 3: ```suggestion
| `packages`     | This object specifies conda or spack package environment information.       |
```

> | `type`         | This represents the type of package builder. you can use `SPACK` or `CONDA` |

Reviewer 3: ```suggestion
| `type`         | This represents the type of package builder. Use `SPACK` or `CONDA`. |
```

> | `packages`     | List of the packages names                                                  |

Reviewer 3: Consistent periods, please.

| `channels`     | List of conda channels, which will be used to download packages             |
| `envFile`      | The package environment file encoded as a base64 string.                    |

> | `mambaImage`   | Name of the docker image to be used in building conda containers            |

Reviewer 3: for building / when buillding

Reviewer 3: Or better: to build

--data '{
    "packages":{
        "type": "CONDA",
        "entries": ["salmon", "bwa"],
        "channels": ["conda-forge", "bioconda"]
    }
}'
```

#### Response

```
{
    "containerToken":"732b73aa17c8",
    "targetImage":"0625dce899da.ngrok.app/wt/732b73aa17c8/hrma017/dev:salmon_bwa--5e49881e6ad74121",
    "expiration":"2024-04-09T21:19:01.715321Z",
    "buildId":"5e49881e6ad74121_1",
    "cached":false,
    "freeze":false
}
```

2. Create singularity image with conda packages

##### Request

```
curl --location 'http://localhost:9090/v1alpha2/container' \
--header 'Content-Type: application/json' \
--data '{
    "format": "sif",
    "containerPlatform": "arm64",
    "packages":{
        "type": "CONDA",
        "entries": ["salmon"],
        "channels": ["conda-forge", "bioconda"]
    },
    "freeze": true,
    "buildRepository": <CONTAINER_REPOSITORY>,
    "towerAccessToken":<YOUR_SEQERA_PLATFORM_TOWER_TOKEN>,
    "towerEndpoint": "http://localhost:8008/api"
}'
```

#### Response

```
{
    "targetImage":"oras://<CONTAINER_REPOSITORY>:salmon--6c084f2e43f86a78",
    "buildId":"6c084f2e43f86a78_1",
    "cached":false,
    "freeze":true
}
```

> Note: You need to add your container registry credentials in seqera platform to use freeze feature which is a requirement for singularity.

Reviewer 3: ```suggestion
:::note
You must add your container registry credentials in Seqera Platform to use the freeze feature. This is a requirement for Singularity.
```

    }
}
```

## Metrics APIs based on SurrealDB

These APIs provide usage (builds and pulls) metrics of Wave per IP and user.

> These APIs required basic authentication, so you need to provide username and password while calling these APIs.

Reviewer 3: ```suggestion
These APIs require basic authentication, so you must provide a username and password while calling these APIs.
```

All Metrics API endpoints have these three common optional query parameters:

| Name      | Description                                                                                   | sample Value |
|-----------|-----------------------------------------------------------------------------------------------|--------------|

> | startDate | Format: `yyyy-mm-dd`, The data of this date will be included in computing the response.       | 2024-02-29   |

Reviewer 3: ```suggestion
| startDate | Format: `yyyy-mm-dd`, This date will be included when computing the response.       | 2024-02-29   |
```

> | endDate   | Format: `yyyy-mm-dd`, The data of this date will be included in computing the response.       | 2024-02-29   |

Reviewer 3: ```suggestion
| endDate   | Format: `yyyy-mm-dd`, This date will be included when computing the response.       | 2024-02-29   |
```

> | limit     | Integer from `0` to `1000` and default is `100`. It limits the number of records in response. | 10           |

Reviewer 3: ```suggestion
| limit     | Integer from `0` to `1000` (default `100`). Limits the number of records in the response. | 10           |
```

> Note: If you only provide startDate then endDate will be today's date. 

Reviewer 3: ```suggestion
:::note
If you only provide `startDate`, `endDate` will be the current date.
:::
```

### Build Metrics APIs

> These APIs are used to get the metrics about the container builds by Wave.

Reviewer 3: ```suggestion
These APIs are used to retrieve metrics about the containers built by Wave.
```

All build metrics API endpoints have one common optional query parameter:

| Name    | Description                                                                                                   | sample Value |
|---------|---------------------------------------------------------------------------------------------------------------|--------------|

> | success | `true` or `false` is allowed and its used to get the metrics of successful container builds performed by Wave | true         |

Reviewer 3: Default value may be more useful than a sample value.

Retrieve the metrics of successful container builds performed by Wave (`true` or `false`).

Below are the different build metrics API endpoints:

### GET `/v1alpha1/metrics/builds`

This endpoint is used to get the total numbers of builds performed by Wave.

### Response

```json
{
    count: integer
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/builds?startDate=2024-02-29&endDate=2024-02-29&success=true"
{"count":2}
```

### GET `/v1alpha1/metrics/builds/ip`

This endpoint is used to get the total numbers of builds per IP performed by Wave.

### Response

```json
{
    result: {
        <IP_ADDRESS>: integer
    }
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/builds/ip?startDate=2024-02-29&endDate=2024-02-29&success=true"
{"result":{"127.0.0.1":2}}
```

### GET `/v1alpha1/metrics/builds/user`

> This endpoint is used to get the total numbers of builds per User performed by Wave.

Reviewer 3: ```suggestion
This endpoint is used to get the total numbers of builds performed by Wave, per user.
```

### Response

```json
{
    result: {
        <EMAIL_ADDRESS>: integer
    }
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/builds/user?startDate=2024-02-29&endDate=2024-02-29&success=true"
{"result":{"test_metrics@seqera.io":2}}
```

### Pull Metrics APIs

These APIs are used to get the metrics about the container pulls through Wave.

All pull metrics API endpoints have one common optional query parameter:

| Name   | Description                                                                                                | sample Value |
|--------|------------------------------------------------------------------------------------------------------------|--------------|

> | fusion | `true` or `false` is allowed and its used to get the metrics of containers with fusion pulled through Wave | true         |

Reviewer 3: Same comment as `success` parameter above.

Below are the different pull metrics API endpoints:

### GET `/v1alpha1/metrics/pulls`

This endpoint is used to get the total numbers of containers pulled through Wave.

### Response

```json
{
    count: integer
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/pulls?startDate=2024-02-29&endDate=2024-02-29&fusion=true"
{"count":2}
```

### GET `/v1alpha1/metrics/pulls/ip`

This endpoint is used to get the total numbers of pulls through Wave per IP.

### Response

```json
{
    result: {
        <IP_ADDRESS>: integer
    }
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/pulls/ip?startDate=2024-02-29&endDate=2024-02-29&success=true"
{"result":{"127.0.0.1":2}}
```

### GET `/v1alpha1/metrics/pulls/user`

This endpoint is used to get the total numbers of pulls through Wave per user.

### Response

```json
{
    result: {
        <EMAIL_ADDRESS>: integer
    }
}
```

#### Example

```
% curl -u foo:bar "http://localhost:9090/v1alpha1/metrics/pulls/user?startDate=2024-02-29&endDate=2024-02-29&fusion=true"
{"result":{"test_metrics@seqera.io":2}}
```

## Metrics APIs based on Redis

> These APIs provide usage (builds and pulls) metrics of Wave for a specific date and/or a specific organisation.

Reviewer 3: ```suggestion
These APIs provide usage (builds and pulls) metrics of Wave for a specific date and/or a specific organization.
```

> These APIs required basic authentication, so you need to provide username and password while calling these APIs.

Reviewer 3: ```suggestion
These APIs require basic authentication, so you must provide a username and password while calling these APIs.
```

> All Metrics API endpoints have these two query parameters, at least one needs to provided to fetch the metrics:

Reviewer 3: ```suggestion
All Metrics API endpoints use these query parameters (at least one is required):
```

| Name | Description                                                           | sample Value |
|------|-----------------------------------------------------------------------|--------------|

> | date | Format: `yyyy-mm-dd`, The data of the required metrics.               | 2024-04-08   |

Reviewer 3: ```suggestion
| date | Format: `yyyy-mm-dd`, The date of the requested metrics.               | 2024-04-08   |
```

> | org  | domain of the organisation used in their emails. e.g. `org=seqera.io` | seqera.io    |

Reviewer 3: ```suggestion
| org  | Domain of the organization used in emails, e.g., `org=seqera.io` | seqera.io    |
```

### Build Metrics API

> These APIs are used to get the metrics about the container builds by Wave.

Reviewer 3: ```suggestion
These APIs are used to retrieve metrics about container builds performed by Wave.
```

### GET `/v1alpha2/metrics/builds`

> This endpoint is used to get the total numbers of builds performed by Wave.

Reviewer 3: ```suggestion
This endpoint is used to retrieve the total number of builds performed by Wave.
```

### Response

```json
{
    count: integer
}
```

#### Examples

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/builds?date=2024-04-08&org=seqera.io"
{"count":4}
```

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/builds?date=2024-04-08"
{"count":6}
```

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/builds?org=seqera.io"
{"count":4}
```
### Pull Metrics API

> These APIs are used to get the metrics about the container pulls through Wave.

Reviewer 3: ```suggestion
These APIs are used to get the metrics about container pulls performed by Wave.
```

### GET `/v1alpha2/metrics/pulls`

> This endpoint is used to get the total numbers of pulls performed through Wave.

Reviewer 3: ```suggestion
This endpoint is used to get the total numbers of pulls performed by Wave.
```

Reviewer 2: its not by wave, but through wave

### Response

```json
{
    count: integer
}
```

#### Examples

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/pulls?date=2024-04-08&org=seqera.io"
{"count":5}
```

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/pulls?date=2024-04-08"
{"count":7}
```

```
% curl -u foo:bar "http://localhost:9090/v1alpha2/metrics/pulls?org=seqera.io"
{"count":5}
```
### Fusion Pull Metrics API

> These APIs are used to get the metrics about the fusion based container pulls through Wave.

Reviewer 3: ```suggestion
These APIs are used to get the metrics about Fusion-based container pulls performed by Wave.
```

### GET `/v1alpha2/metrics/fusion/pulls`

> This endpoint is used to get the total numbers of pulls of fusion based containers performed through Wave.

Reviewer 3: ```suggestion
This endpoint is used to get the total number of Fusion-based container pulls performed by Wave.
```
